ENH: special: add more functions to `_support_alternative_backends`

NeilGirdhar opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->

#### What does this implement/fix?
<!--Please explain your changes.-->

#### Additional information
<!--Any additional information you think is important.-->

NeilGirdhar committed:
Add Array API versions of special functions

NeilGirdhar commented:
Based on https://github.com/data-apis/array-api/issues/725#issuecomment-2775874787

mdhaber commented:
@crusaderky should this wait for gh-22718 or gh-22639, or would it be helpful to have more functions to test those with?

mdhaber commented on scipy/special/_support_alternative_backends.py:
```diff
@@ -19,6 +20,12 @@ def get_array_special_func(f_name, xp, n_array_args):
+        if f_name in {'gamma', 'betaln', 'beta', 'betainc', 'expi', 'expn'}:
+            raise NotImplementedError
+        # These functions have incompatible interaces.
+        if f_name in {'logsumex'}
```
```suggestion
        if f_name in {'logsumexp'}
```
But this can just be removed. `scipy.special.logsumexp` has a generic implementation, and it does not go through this machinery.

NeilGirdhar replied:
Oh, so it will call through the other Array API functions?

mdhaber replied:
It is implemented in terms of array API standard calls.
https://github.com/scipy/scipy/blob/25ca572f6dcaf2d7ad38d96fdc9def33b2f8a385/scipy/special/_logsumexp.py#L13

NeilGirdhar replied:
Awesome!

crusaderky commented on scipy/special/_support_alternative_backends.py:
```diff
@@ -51,22 +55,38 @@ def wrapped(*args, **kwargs):
     'erfc': 1,
+    'erfinv': 1,
+    'zeta': 2,
+    # No one implements binom yet.
+    # 'binom': 2,
```
```suggestion
    'binom': 2,
```
Dask and MArray implement _all elementwise ufuncs_ when they wrap around numpy!

NeilGirdhar replied:
Great!

If we expose `is_dask_namespace`, I can test for that and let `binom` through for that case.

How do I test for MArray?

crusaderky replied:
> Great!
> 
> If we expose `is_dask_namespace`, I can test for that and let `binom` through for that case.

Please merge from main, there's news for you :wink: #22639

> How do I test for MArray?

At the moment, specific special functions that are needed by scipy.stats are indirected tested in there.
For a more systematic test effort, I would advise to talk it through thoroughly with @mdhaber and write a dedicated PR for it.

crusaderky replied:
`binom` is not a function; it's a namespace for distributions (`binoms.ppf` etc.). This delegation system is not fit for purpose for it.

NeilGirdhar replied:
Sorry, I'm confused here.  `scipy.stats.binom` is a function:
```python
[utM] In [10]: ss.binom(5, 2)
Out[10]: np.float64(10.0)
```

crusaderky replied:
Apologies, I got confused with `scipy.stats.binom`

crusaderky commented on scipy/special/_support_alternative_backends.py:
```diff
@@ -19,6 +20,9 @@ def get_array_special_func(f_name, xp, n_array_args):
     if is_numpy(xp):
         f = getattr(_ufuncs, f_name, None)
     elif is_torch(xp):
+        # These functions are not implemented.
+        if f_name in {'gamma', 'betaln', 'beta', 'betainc', 'expi', 'expn'}:
+            raise NotImplementedError
```
```suggestion
```
I expect them to work on CPU through the numpy round-trip a few lines below.
On GPU, they should be skipped in the test (there is already a section for it).

NeilGirdhar replied:
Do we have a way to test for PyTorch-cpu?  If not, does it makes more sense to block them on PyTorch since it would lead to the odd behavior whereby a program works just fine on PyTorch-cpu, but fails on PyTorch-gpu?

crusaderky replied:
```bash
python dev.py test -b torch  # on CPU float64
SCIPY_DEVICE=cuda python dev.py test -b torch  # on GPU float64
SCIPY_DEFAULT_DTYPE=float32 python dev.py test -b torch  # on CPU float32
```
CI runs all three of them.

>  a program works just fine on PyTorch-cpu, but fails on PyTorch-gpu?

We already have tons of stuff that works fine on torch/jax CPU but fails on GPU. Basically anything with a Cython or C internal call.

NeilGirdhar replied:
Sure, but it seems that you're asking me to let these functions pass through here?  Do you want me to unconditionally let them pass through or just when PyTorch is running on the CPU?

crusaderky replied:
unconditionally. They will then crash and burn when `np.asarray` tries to implicitly move the GPU buffer to host memory a few lines later.

crusaderky commented:
> @crusaderky should this wait for [gh-22718](https://github.com/scipy/scipy/pull/22718) or [gh-22639](https://github.com/scipy/scipy/pull/22639), or would it be helpful to have more functions to test those with?

merge order doesn't matter

crusaderky commented:
None of these changes will be picked up (and the unit tests should definitely tell you) unless you also change `scipy/special/__init__.py:787:791`.
I suggest to just change it to 
```python
from ._support_alternative_backends import *
```

crusaderky commented on scipy/special/_support_alternative_backends.py:
```diff
@@ -51,22 +55,38 @@ def wrapped(*args, **kwargs):
+    # 'binom': 2,
+    'expi': 1,
+    'expn': 2,
+    'softmax': 1,
```
This function is not in `_ufuncs.py`. Even if you were to fix the import, it is not elementwise and cannot use this delegation system without structural changes to it. 

This delegation system can't realise that the function is already API agnostic, so it will try to convert everything to numpy before pushing data through it. On Dask, this will cause reductions to be applied locally to chunks instead of globally. On MArray, it will incorrectly calculate the mask.

I expect (hope) that the Dask test will fail  really loudly saying that the numbers coming out of it are wrong. Note that there is no test for MArray.

crusaderky replied:
Please don't overload this module, which was never meant for anything that is not an elementwise ufunc and is not fit for purpose for it.
I'm very happy to go through a more structural redesign and unification of the delegation system, but it should be its own separate design effort and PR. This is out of scope here.

NeilGirdhar replied:
Removed `softmax`, and `log_softmax`

NeilGirdhar commented:
@crusaderky Thanks for your thorough and quick review!  My apologies that I'm a bit new to this code, so thanks for being patient.  I'll answer some of your comments now to try and gather some information before addressing your concerns.

NeilGirdhar commented:
I looked and couldn't find the justification for vendoring array-api-compat.  My issue is that the code here uses a very old version of compat and very awkward tests.  In the current version of compat, we would use:
```
    is_numpy_namespace
    is_cupy_namespace
    is_torch_namespace
    is_ndonnx_namespace
    is_dask_namespace
    is_jax_namespace
    is_pydata_sparse_namespace
    is_array_api_strict_namespace
```
It's unfortunate that we can't use these functions as they would lead to more readable code.

crusaderky commented:
> I looked and couldn't find the justification for vendoring array-api-compat. My issue is that the code here uses a very old version of compat and very awkward tests. In the current version of compat, we would use:
> 
> ```
>     is_numpy_namespace
>     is_cupy_namespace
>     is_torch_namespace
>     is_ndonnx_namespace
>     is_dask_namespace
>     is_jax_namespace
>     is_pydata_sparse_namespace
>     is_array_api_strict_namespace
> ```
> 
> It's unfortunate that we can't use these functions as they would lead to more readable code.

We're currently vendoring the version released 2 weeks ago and we can bump to git tip if needed.
Those functions are imported through `scipy._lib._array_api`, where they're renamed to `is_numpy` etc.

NeilGirdhar commented:
Got it.  (My mistake about it being old).  Why weren't the original function names kept?

lucascolley commented:
> Why weren't the original function names kept?

just a historical artifact from when we were using much less of this stuff. Feel free to send a refactor PR if that would help avoid confusion.

crusaderky commented on scipy/special/tests/test_support_alternative_backends.py:
```diff
@@ -56,15 +56,22 @@ def test_rel_entr_generic(dtype):
 @pytest.mark.parametrize('shapes', [[(0,)]*4, [tuple()]*4, [(10,)]*4,
                                     [(10,), (11, 1), (12, 1, 1), (13, 1, 1, 1)]])
 def test_support_alternative_backends(xp, f_name, n_args, dtype, shapes):
+    if f_name in {'expn', 'polygamma'}:
+        pytest.skip(f"`{f_name}` has limited domain.")
+    if f_name in {'multigammaln'}:
+        pytest.skip(f"`{f_name}` requires some integer arguments.")
```
It would be _very_ good to get https://github.com/scipy/scipy/pull/22718 in so that we can think how to build some tooling for these on top of it.

NeilGirdhar replied:
Agreed.  Should I wait for that PR?

crusaderky replied:
I think it's ok to leave these untested for the time being. Just know that the more changes you do to the test module the more either of us will have to rework to fix a conflict.

NeilGirdhar commented:
> Could you clarify your question here?

Sorry, I'm just frustrated with my inability to build scipy (which is making finishing this feature take longer than I anticipated).  I just thought that Jax has Jaxlib (which I assume has all the compiled code), and this partition makes Jax easier to develop since it can use ordinary setuptools (or whatever) rather than Meson.

lucascolley commented:
> I just thought that Jax has Jaxlib (which I assume has all the compiled code), and this partition makes Jax easier to develop since it can use ordinary setuptools (or whatever) rather than Meson.

Maybe one day :) but the reality is that compiled code is scattered all over SciPy in various languages and this is a fast moving area, with various things getting translated and rewritten etc.

`xsf` is an example taking a step in this direction, where part of the compiled code backend will eventually be broken out into a separate library (of course we might still vendor that in the repo rather than distributing a separate dependency for the time-being). But it will be an enormous amount of work to do this for every corner of SciPy, so I wouldn't keep your hopes up for it happening soon :sweat_smile:.

> I'm just frustrated with my inability to build scipy

Don't worry, this a common experience. Seriously though, this is why https://github.com/rgommers/pixi-dev-scipystack will become very valuable.

NeilGirdhar commented:
Awesome thanks for explaining, and thanks for your help!

NeilGirdhar commented:
Sorry that I haven't had time to work on this.  Has this been implemented now by someone else?  Should we close?

crusaderky commented:
It hasn't; however the module has changed quite a lot in the meantime.

steppi commented:
I can clean up the merge conflicts and will take a close look at everything.

steppi committed:
Merge branch 'main' into array_api

steppi committed:
MAINT/TST: Reintroduce support for betaln + update tests in response

steppi committed:
MAINT: Re-add binom

steppi committed:
MAINT: Add psi-function and necessary machinery

steppi committed:
MAINT: Add back erfinv

steppi committed:
MAINT: Add back expi

steppi committed:
MAINT/TST: Add back expn and test machinery for int only params

steppi committed:
MAINT: Add back gamma

steppi committed:
MAINT/TST: Add back polygamma and necessary test infra

steppi committed:
TST: Make positive_only configurable in _FuncInfo

steppi committed:
MAINT/TST: Add back zeta and relevant skip logic

steppi committed:
ENH: Allow multigammaln to take 0d arrays for d

steppi committed:
MAINT/TST: Add multigammaln and fix most issues with testing

steppi committed:
TST: Skip alt backend multigammaln tests

steppi committed:
ENH: Add bdtr cupy support

steppi committed:
ENH: Add cupy support for bdtrc

steppi committed:
ENH: Add cupy support for bdtri

steppi committed:
ENH: Add cupy support for betaincinv

steppi committed:
ENH: Add cupy support for boxcox

steppi committed:
ENH: Add cupy support for boxcox1p

steppi committed:
ENH: Add CuPy support for cbrt

steppi committed:
ENH: Add CuPy support for chdtri

steppi committed:
ENH: Add CuPy support for cosdg

steppi committed:
ENH: Add CuPy support for cosm1 and cotdg

steppi committed:
ENH: Add CuPy support for ellipk and ellipkm1

steppi committed:
ENH: Sure up cupy support in special

steppi committed:
ENH: Sure up torch support in special

steppi committed:
ENH: Sure up Jax support in special

steppi committed:
STY: make code style more consistent

steppi committed:
TST: don't test large ints for poch on jax

steppi committed:
TST: Respect test_large_ints in mismatched_dtypes tests

steppi committed:
Merge branch 'main' into array_api_special_merge

steppi commented:
I've resolved the merge conflicts and carried from there to add dispatch to all available special functions across backends. I'll post a comment explaining everything tomorrow.

steppi commented:
I'll investigate the failures and try to figure out why they aren't happening locally.

ev-br commented on scipy/special/tests/test_support_alternative_backends.py:
```diff
@@ -21,41 +22,92 @@
+def _to_python_scalar(x):
+    xp = array_namespace(x)
+    if x.ndim != 0:
+        x = x.take(0)
```
`.take` method is not array API compliant, I think?

Also it's a bit tricky to follow the logic --- or I am just being dense --- if `x` can be a scalar, could you write it roughly

```
if isinstance(x, bool |int | float | complex):
     return x 

xp = array_namespace(x)

if xp.isdtype(x.dtype, ....):
  ....
```

steppi replied:
https://data-apis.org/array-api/latest/API_specification/generated/array_api.take.html#take

I didn't spend any time thinking about writing this in a nice way, so can definitely be simplified.

lucascolley replied:
I think Evgeni means that we should be using `xp.take` rather than `x.take`.

steppi replied:
Oh, ha. Sorry, I didn't notice I didn't do that

steppi commented on scipy/special/tests/test_support_alternative_backends.py:
```diff
@@ -21,41 +22,92 @@
 
 
-def _skip_or_tweak_alternative_backends(xp, f_name, dtypes):
+def _get_native_namespace_name(xp):
+    if is_cupy(xp):
+        return "cupy"
+    if is_dask(xp):
+        return "dask.array"
+    if is_jax(xp):
+        return "jax.numpy"
+    if is_numpy(xp):
+        return "numpy"
+    if is_torch(xp):
+        return "torch"
+    if is_array_api_strict(xp):
+        return "array_api_strict"
+    raise ValueError(f"Received unsupported array namspace {xp}")
```
Yeah, `enum` of backends sounds like a good idea.

steppi committed:
MAINT: Get torch tests passing

steppi committed:
MAINT: Get rid of unnecessary and hastily written helper function

steppi commented:
The JAX GPU tests are failing because the JAX in CI is `v0.6.0`, but `hyp2f1` landed in `v0.6.1`. For whatever reason, `v0.6.1` and `v0.6.2` are up on PyPI but haven't landed in the conda feedstock yet, https://github.com/conda-forge/jax-feedstock/pulls. I can just remote it for now. Having a `requires_version` option in `_FuncInfo` seems like it would be useful though, but leaving out for now.

There's also some unexpected failures to pickle for random functions. I'll keep looking into that.

steppi committed:
STY: Remove unused imports

steppi committed:
MAINT: clarify by using more descriptive variable name

steppi committed:
MAINT: Comment out hyp2f1 since not available in jax in CI yet

steppi committed:
STY: Remove erroneous formatting changes

steppi commented:
```
=================================== FAILURES ===================================
______________________________ test_pickle[cbrt] _______________________________

func = <ufunc 'cbrt'>

    @pytest.mark.parametrize("func", [nfo.wrapper for nfo in _special_funcs])
    def test_pickle(func):
>       roundtrip = pickle.loads(pickle.dumps(func))
                                 ^^^^^^^^^^^^^^^^^^

func       = <ufunc 'cbrt'>

scipy/special/tests/test_support_alternative_backends.py:294: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/numpy/core/__init__.py:32: in __getattr__
    _raise_warning(attr_name)
        attr       = <ufunc 'cbrt'>
        attr_name  = 'cbrt'
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

attr = 'cbrt', submodule = None

    def _raise_warning(attr: str, submodule: str | None = None) -> None:
        new_module = "numpy._core"
        old_module = "numpy.core"
        if submodule is not None:
            new_module = f"{new_module}.{submodule}"
            old_module = f"{old_module}.{submodule}"
>       warnings.warn(
            f"{old_module} is deprecated and has been renamed to {new_module}. "
            "The numpy._core namespace contains private NumPy internals and its "
            "use is discouraged, as NumPy internals can change without warning in "
            "any release. In practice, most real-world usage of numpy.core is to "
            "access functionality in the public NumPy API. If that is the case, "
            "use the public NumPy API. If not, you are using NumPy internals. "
            "If you would still like to access an internal attribute, "
            f"use {new_module}.{attr}.",
            DeprecationWarning,
            stacklevel=3
        )
E       DeprecationWarning: numpy.core is deprecated and has been renamed to numpy._core. The numpy._core namespace contains private NumPy internals and its use is discouraged, as NumPy internals can change without warning in any release. In practice, most real-world usage of numpy.core is to access functionality in the public NumPy API. If that is the case, use the public NumPy API. If not, you are using NumPy internals. If you would still like to access an internal attribute, use numpy._core.cbrt.

attr       = 'cbrt'
new_module = 'numpy._core'
old_module = 'numpy.core'
submodule  = None

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/numpy/core/_utils.py:10: DeprecationWarning
______________________________ test_pickle[exp2] _______________________________

func = <ufunc 'exp2'>

    @pytest.mark.parametrize("func", [nfo.wrapper for nfo in _special_funcs])
    def test_pickle(func):
>       roundtrip = pickle.loads(pickle.dumps(func))
                                 ^^^^^^^^^^^^^^^^^^

func       = <ufunc 'exp2'>

scipy/special/tests/test_support_alternative_backends.py:294: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/numpy/core/__init__.py:32: in __getattr__
    _raise_warning(attr_name)
        attr       = <ufunc 'exp2'>
        attr_name  = 'exp2'
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

attr = 'exp2', submodule = None

    def _raise_warning(attr: str, submodule: str | None = None) -> None:
        new_module = "numpy._core"
        old_module = "numpy.core"
        if submodule is not None:
            new_module = f"{new_module}.{submodule}"
            old_module = f"{old_module}.{submodule}"
>       warnings.warn(
            f"{old_module} is deprecated and has been renamed to {new_module}. "
            "The numpy._core namespace contains private NumPy internals and its "
            "use is discouraged, as NumPy internals can change without warning in "
            "any release. In practice, most real-world usage of numpy.core is to "
            "access functionality in the public NumPy API. If that is the case, "
            "use the public NumPy API. If not, you are using NumPy internals. "
            "If you would still like to access an internal attribute, "
            f"use {new_module}.{attr}.",
            DeprecationWarning,
            stacklevel=3
        )
E       DeprecationWarning: numpy.core is deprecated and has been renamed to numpy._core. The numpy._core namespace contains private NumPy internals and its use is discouraged, as NumPy internals can change without warning in any release. In practice, most real-world usage of numpy.core is to access functionality in the public NumPy API. If that is the case, use the public NumPy API. If not, you are using NumPy internals. If you would still like to access an internal attribute, use numpy._core.exp2.

attr       = 'exp2'
new_module = 'numpy._core'
old_module = 'numpy.core'
submodule  = None

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/numpy/core/_utils.py:10: DeprecationWarning
=========================== short test summary info ============================
FAILED scipy/special/tests/test_support_alternative_backends.py::test_pickle[cbrt] - DeprecationWarning: numpy.core is deprecated and has been renamed to numpy._core. The numpy._core namespace contains private NumPy internals and its use is discouraged, as NumPy internals can change without warning in any release. In practice, most real-world usage of numpy.core is to access functionality in the public NumPy API. If that is the case, use the public NumPy API. If not, you are using NumPy internals. If you would still like to access an internal attribute, use numpy._core.cbrt.
FAILED scipy/special/tests/test_support_alternative_backends.py::test_pickle[exp2] - DeprecationWarning: numpy.core is deprecated and has been renamed to numpy._core. The numpy._core namespace contains private NumPy internals and its use is discouraged, as NumPy internals can change without warning in any release. In practice, most real-world usage of numpy.core is to access functionality in the public NumPy API. If that is the case, use the public NumPy API. If not, you are using NumPy internals. If you would still like to access an internal attribute, use numpy._core.exp2.
= 2 failed, 66205 passed, 7604 skipped, 11691 deselected, 213 xfailed, 11 xpassed in 274.70s (0:04:34) =
```

so strange. I can't reproduce this on my Linux machine or my Mac mini. The only commonality is that `exp2` and `cbrt` are ufuncs which are also in NumPy, (though SciPy has its own implementations from Cephes). For now, I guess I'll just suppress the warning and make an issue about it.

steppi committed:
TST: Ignore mysterious deprecation warning in pickle tests

steppi committed:
MAINT: Fix garbled explanation string

lucascolley reviewed (approved):
I think I'm happy for this to go in when CI is green!

steppi commented:
Thanks @lucascolley. It looks like the failures are unrelated so far. A few comments I wanted to add.

1. I didn't add `special.sinc` because it isn't in `jax.scipy.special` but instead is in `jax.numpy` (since `sinc` is in NumPy and SciPy's `sinc` is just an alias for NumPy's). But `sinc` is in CuPy and PyTorch, I just had missed that only the JAX tests were failing. I made a determination to not add any more special case logic, but thought to follow up with allowing functions to come from an alternative namespace. Since `sinc` is in PyTorch and CuPy though, I made as well at it for those in this PR.

2. I wanted to make sure that adding all of these extra attributes to the `_FuncInfo` wouldn't lead to bloat, so measured with [pympler](https://pympler.readthedocs.io/en/latest/). Things don't look too bad.

Memory usage on main after importing `special`.

```
                         types |   # objects |   total size
============================== | =========== | ============
                           str |       28758 |      5.27 MB
                          code |        9387 |      3.79 MB
                          dict |        7840 |      2.65 MB
                          type |        1311 |      1.38 MB
                         tuple |       10553 |    639.02 KB
  Cython.Shadow.CythonMetaType |         198 |    329.48 KB
                          list |        2439 |    285.39 KB
                           set |         266 |    245.11 KB
                   abc.ABCMeta |         131 |    184.99 KB
            wrapper_descriptor |        2421 |    170.23 KB
         weakref.ReferenceType |        1997 |    156.02 KB
    builtin_function_or_method |        1933 |    135.91 KB
                   numpy.ufunc |         505 |    126.25 KB
             method_descriptor |        1354 |     95.20 KB
                     frozenset |         257 |     94.21 KB
```

Memory usage on this PR branch after importing `special`.

```
                         types |   # objects |   total size
============================== | =========== | ============
                           str |       28839 |      5.29 MB
                          code |        9393 |      3.80 MB
                          dict |        7990 |      2.69 MB
                          type |        1311 |      1.38 MB
                         tuple |       10635 |    644.56 KB
  Cython.Shadow.CythonMetaType |         198 |    329.48 KB
                          list |        2934 |    314.41 KB
                           set |         261 |    244.05 KB
                   abc.ABCMeta |         131 |    184.99 KB
            wrapper_descriptor |        2421 |    170.23 KB
         weakref.ReferenceType |        1987 |    155.23 KB
    builtin_function_or_method |        1927 |    135.49 KB
                   numpy.ufunc |         506 |    126.50 KB
             method_descriptor |        1354 |     95.20 KB
                     frozenset |         257 |     94.21 KB
```

3. Some of the weird edge cases, like `multigammaln` having a parameter which can only be a Python int should probably be changed on the function side, rather than maintaining support for handling all of these edge cases.

steppi committed:
ENH: Add alt backend support for sinc

steppi committed:
MAINT: avoid converting integer only types to float in test

steppi committed:
TST: change argtypes to int_only and fix broken test code

steppi commented on scipy/special/tests/test_support_alternative_backends.py:
```diff
@@ -67,37 +87,101 @@ def _skip_or_tweak_alternative_backends(xp, f_name, dtypes):
+        # to float64 instead. Do not convert integer only args to float32 though.
+        dtypes_np_ref = [
+            'float32' if 'int' in dtype and not needs_int else dtype
+            for dtype, needs_int in zip(dtypes, int_only)
```
I saw that's the problem in the array api / torch etc tests. I wonder if the strange failures in the Linux / 32 bit tests are for the same reason though

steppi committed:
TST: Ensure int_only is tuple of bool of proper size in tests

steppi committed:
MAINT: Ensure digamma alias gets alt backend support

steppi committed:
MAINT: properly annotate lack of jax_jit support

steppi committed:
MAINT: Fix dtype related failures

steppi commented:
OK, I think that should take care of the remaining test failures.

steppi commented on scipy/special/tests/test_support_alternative_backends.py:
```diff
@@ -21,83 +21,167 @@
-        # numpy arrays to float32 to prevent them from being automatically promoted
-        # to float64 instead.
-        dtypes_np_ref = ['float32' if 'int' in dtype else dtype for dtype in dtypes]
+    if (is_torch(xp) and xp_default_dtype(xp) == xp.float32):
+        # On PyTorch with float32 default dtype, all ints are promoted
+        # to float32, but when falling back to NumPy/SciPy int64 is promoted
+        # instead to float64. Integer only parameters essentially do not
+        # participate in determination of the result type in PyTorch with
+        # float32 default dtype, but will impact the output dtype as if
+        # they were float64 when falling back to NumPy/SciPy.
+        if not nfo.torch_native:
+            pytest.xfail("dtypes no not match")
```
The previous comment made it seem like PyTorch randomly mapped to `float32` or `float64` in this case, but I discovered `float64` output only occurs when PyTorch falls back to NumPy/SciPy, which have no concept of `default_dtype`. I added a flag to the `_FuncInfo` objects to encode whether PyTorch falls back, and just insert the `xfails` here.

lucascolley replied:
please can you investigate getting rid of `xp_default_dtype` and replacing it with `xpx.default_dtype`? https://data-apis.org/array-api-extra/generated/array_api_extra.default_dtype.html

steppi commented on scipy/special/tests/test_support_alternative_backends.py:
```diff
@@ -21,83 +21,167 @@
-        # When they are promoted to float32, explicitly convert the reference
-        # numpy arrays to float32 to prevent them from being automatically promoted
-        # to float64 instead.
-        dtypes_np_ref = ['float32' if 'int' in dtype else dtype for dtype in dtypes]
```
This handled the different promotion rules by modifying the inputs, but doesn't work for instance when there are integer only arguments. I changed it to modify the output to the correct type instead, directly at the point where the reference output is computed.

steppi commented on scipy/special/tests/test_support_alternative_backends.py:
```diff
@@ -108,10 +192,25 @@ def test_support_alternative_backends(xp, func, nfo, dtype, shapes):
     res = nfo.wrapper(*args_xp)  # Also wrapped by lazy_xp_function
     ref = nfo.func(*args_np)  # Unwrapped ufunc
-
+    if (
+            is_torch(xp)
+            and xp_default_dtype(xp) == xp.float32
+            and "float64" not in dtypes
+    ):
+        # int64 promotes like float32 on torch with default dtype = float32
+        # cast reference if needed
+        ref = np.float32(ref)
```
Eventually we should remove code duplication rather than just pasting this line in everywhere a reference output is computed, but I'd prefer to punt on what the best way to do this. Using duplication to postpone the decision basically.

steppi reviewed (commented):
Added a few comments explaining what I did. I also realized I wasn't annotating functions which rely on NumPy/SciPy fallback for JAX correctly. I updated those to use `jax_jit=False`.

steppi commented:
Oh, looks like `test_public_api.py` needs to be updated for how I handled giving alternative backend support for the alias `digamma` for `psi`.

steppi committed:
MAINT: make digamma work with deprecated direct import from basic

steppi commented:
It looks like the weird memory error on 32 bit Linux still happens even after everything else is fixed. I'll see if I can reproduce in a 32 bit Linux Docker container.

steppi committed:
MAINT: Set jax_jit=False for bdtr

steppi commented:
> It looks like the weird memory error on 32 bit Linux still happens even after everything else is fixed. I'll see if I can reproduce in a 32 bit Linux Docker container.
> 
> Update: I can reproduce it locally now. Will investigate further tomorrow.

Ah, it's a confluence of things. On 32 bit systems, the NumPy default `int` is `int32`, the tests use `int64` type. The overloads for some functions look like this

```python
    "bdtr": {
        "_legacy.pxd": {
            "bdtr_unsafe": "ddd->d"
        },
        "xsf_wrappers.h": {
            "cephes_bdtr_wrap": "dpd->d"
        }
    },
```

the `p` in `"dpd->d"` is for `ptrdiff_t` which will always be the NumPy default `int`. This `int32` on 32 bit systems, so when you pass an `int64`, the deprecated "`ddd->d`" signature is chosen, which leads to deprecation warnings. A lot of deprecation warnings. So many deprecation warnings that the limited 4GB of RAM you get with a 32 bit address space gets flooded when pytest is doing its teardown and we see an out of memory error.

For a principled fix, maybe we should provide the `int32` and `int64` overload for all integer parameters, not just an overload for the NumPy default `int`. Sebastian Berg had mentioned this here, https://github.com/scipy/scipy/pull/21401#issuecomment-2296164860, using a `long long` `q` typecode, as a way to allow passing `int64` on 32 bit platforms. For now, I can just conditionally make the integer types `int32` on 32 bit systems in the tests.

steppi committed:
TST: Use np default int in alt backend special function tests

steppi committed:
TST: ensure numpy default int used in special alt backend tests

steppi commented:
@lucascolley, it looks like I finally got all tests passing.

lucascolley commented on scipy/special/tests/test_support_alternative_backends.py:
```diff
@@ -21,83 +21,172 @@
+        # float32 default dtype, but will impact the output dtype as if
+        # they were float64 when falling back to NumPy/SciPy.
+        if not nfo.torch_native:
+            pytest.xfail("dtypes no not match")
```
```suggestion
            pytest.xfail("dtypes do not match")
```

lucascolley commented on scipy/special/_support_alternative_backends.py:
```diff
@@ -256,40 +299,577 @@ def fun(t, df, p):  return stdtr(df, t) - p
     _FuncInfo(_ufuncs.chdtr, 2, generic_impl=_chdtr),
-    _FuncInfo(_ufuncs.chdtrc, 2, generic_impl=_chdtrc),
+    _FuncInfo(_ufuncs.chdtrc, 2, generic_impl=_chdtrc,
+              # gh-20972
```
totally optional nit but just for future reference, this makes a nice hyperlink in the GitHub web UI

```suggestion
              # scipy/scipy#20972
```

lucascolley reviewed (changes requested):
thanks @steppi , one request

steppi committed:
Apply suggestions from code review [skip ci]

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

steppi committed:
MAINT: change xp_default_dtype to xpx.default_dtype

lucascolley commented:
thanks! You said:

> The previous comment made it seem like PyTorch randomly mapped to float32 or float64 in this case, but I discovered float64 output only occurs when PyTorch falls back to NumPy/SciPy, which have no concept of `default_dtype`.

Can you check whether that last phrase still holds true now that we are using `xpx.default_dtype`? If it doesn't, are there any changes to be made?

lucascolley reviewed (approved):
unless my previous comment prompts any further thoughts, happy to click the green button. Thanks all!

steppi commented:
> thanks! You said:
> 
> > The previous comment made it seem like PyTorch randomly mapped to float32 or float64 in this case, but I discovered float64 output only occurs when PyTorch falls back to NumPy/SciPy, which have no concept of `default_dtype`.
> 
> Can you check whether that last phrase still holds true now that we are using `xpx.default_dtype`? If it doesn't, are there any changes to be made?
> 
> EDIT: thinking a bit more, I guess there won't be any more work to do for this specific case, as there is still no variable default dtype for NumPy. But maybe this affects something on the JAX side of things?

Yeah, nothing should change here due to using `xpx.default_dtype`. Like you wrote in the edit, NumPy has no concept of default dtype, so computations going through NumPy can't match PyTorch's  `default_dtype=float32` promotion behavior. I'm not sure if we have JAX tests with the default dtype set to `float32`, but in any case, when `SCIPY_DEFAULT_DTYPE="float32"` then in JAX, `int64` isn't even an available type, since what happens when `SCIPY_DEFAULT_DTYPE="float32"` is that `JAX_ENABLE_X64` gets set to `False`. I don't think JAX has the same notion of default dtype that PyTorch has, instead you can turn `X64` support off and on. This means that the problems that arise when PyTorch falls back to SciPy/NumPy cannot occur, since it's not possible to mix `float32` and `int64` in the first place.

lucascolley commented:
> I'm not sure if we have JAX tests with the default dtype set to float32

Indeed we don't, AFAIK since JAX basically has two modes: either default `float32` and throw errors when trying to use double precision, or default to double precision. Unless anything has moved recently, I don't think there is a mode which enables double precision while still allowing you to default to `float32`. So I don't think these tests are possible atm.

---

gh-23081 next :) ?

steppi commented:
> [gh-23081](https://github.com/scipy/scipy/pull/23081) next :) ?

Yes, working on finishing it now. Main thing left is a comment explaining what is and isn't included in the tables and why some submodules are missing.
